### ISSUE TITLE
ToolsetRegistry is not persistent enough

### ISSUE THREAD

**Issue by mauritsvanrees** [zopefoundation/Products.GenericSetup#86]
I did a PR #26 more than three years ago, which was released in GenericSetup 1.8.3. It made it possible to move a tool from the required list to the forbidden list, or the other way around.

I now see a problem with that. I have a `toolset.xml` which does exactly that, and it seems to work, but its resulting changes are not persisted: they are gone after a restart. If more changes are done to `portal_setup` before a restart, then things might be fine. But not when there is an immediate restart.

I can work on a PR for that.

Let me describe the situation:
- client code with `Tool` class inheriting from `CMFCore.utils.UniqueObject/ImmutableId` and id "correct_id".
- `toolset = portal_setup.getToolsetRegistry()`
- `toolset.listForbiddenTools()` is an empty list.
- `toolset.listRequiredToolInfo()` dictionary has one Tool class with `correct_id` and one with `typo_id`.
- When I look in the git history of the project, this situation has existed since at least 2012.
- A few months ago I was tired of this `typo_id` and I manually removed this object in the ZMI.
- Result now: any profile or upgrade with a `toolset.xml` causes GenericSetup to try to create the missing `typo_id` but this fails.

I wanted to fix that. I actually already had a correct `toolset.xml` which should fix it:

```
<?xml version="1.0"?>
<tool-setup>
  <forbidden tool_id="typo_id"/>
  <required tool_id="correct_id"
            class="client.code.Tool"/>
</tool-setup>
```

Applying this initially works:
- the forbidden tools list contains `typo_id`
- the required dictionary only has my class Tool with `correct_id`.

But after a restart, the old situation is there again, and any other `toolset.xml` that is applied, throws an error, just like before.
The reason is that the forbidden and required information is stored in a [simple list and dictionary](https://github.com/zopefoundation/Products.GenericSetup/blob/2.0.1/Products/GenericSetup/registry.py#L650-L653) instead of in persistent data.
In my case this is in GenericSetup 1.8.9 on Plone 4.3.18, but I see it is the same on the master branch.

I have tried a few things:
- Set `toolset._p_changed = True`: does not help.
- Add `self._forbidden = self._forbidden` whenever changing `self._forbidden`. Same for `self._required`. Does not help.
- Set `portal_setup._p_changed = True`: helps.

I expected that one of the first two would help, because [`ToolsetRegistry` inherits from `Acquisition.Implicit`](https://github.com/zopefoundation/Products.GenericSetup/blob/2.0.1/Products/GenericSetup/registry.py#L554). But that is maybe only when it inherits from `Explicit`.

The question is then where to fix it.

1. If we try to fix it in the `ToolsetRegistry` class, then we would need to call something like `self.__parent__._p_changed = True` which seems wrong: it accesses a private variable of the parent.
2. We could change it in the `SetupTool` class: after calling [`toolset.parseXML`](https://github.com/zopefoundation/Products.GenericSetup/blob/2.0.1/Products/GenericSetup/tool.py#L118) we can set `self._p_changed` to True. It might need to be done in more cases, but at least this is the main one.
3. Possibly we could change the list and dictionary to their persistent counterparts on the fly, but that seems more invasive, with more chance of introducing bugs.

I think I will go with number 2.

**Comment by ale-rt** [comment]
Doing `self._forbidden = self._forbidden[:]` helps?
Probably not, but if it is easy for you to try I would give it a spin.
About the choice of option number 2., it seems a little bit unfortunate that we mark the object changed even if nothing changes (at least this is what I get from reading the patch).
Anyway you explained very well the pros and cons of all the options :)

**Comment by mauritsvanrees** [comment]
The fix was already released in 2.0.2 in January.
I have now made a release of the 1.8 branch with this: 1.8.11.
